[spmd expansion] support scalar_tensor

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98233
* #98232
* __->__ #98231
* #98230
* #98229

scalar_tensor is a pure factory function that can't be handled by DTensor prop rule and needs to be currently handled in spmd expansion.

Differential Revision: [D44615039](https://our.internmc.facebook.com/intern/diff/D44615039/)

Commit:
[spmd expansion] support scalar_tensor

scalar_tensor is a pure factory function that can't be handled by DTensor prop rule and needs to be currently handled in spmd expansion.

Differential Revision: [D44615039](https://our.internmc.facebook.com/intern/diff/D44615039/)

[ghstack-poisoned]

mrshenli reviewed (approved):
this one supersedes #98179

Commit:
Update on "[spmd expansion] support scalar_tensor"


scalar_tensor is a pure factory function that can't be handled by DTensor prop rule and needs to be currently handled in spmd expansion.

Differential Revision: [D44615039](https://our.internmc.facebook.com/intern/diff/D44615039/)

[ghstack-poisoned]

Commit:
Update on "[spmd expansion] support scalar_tensor"


scalar_tensor is a pure factory function that can't be handled by DTensor prop rule and needs to be currently handled in spmd expansion.

Differential Revision: [D44615039](https://our.internmc.facebook.com/intern/diff/D44615039/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/distribute.py:
```diff
@@ -507,11 +507,24 @@ def _convert_to_distributed(
+                        device=node.kwargs["device"]
+                    ),
+                    schemas[0].mesh,
+                    [Replicate()],
```
I assume this won't trigger collective in backward pass, as it's `requires_grad=False`?

Commit:
Update on "[spmd expansion] support scalar_tensor"


scalar_tensor is a pure factory function that can't be handled by DTensor prop rule and needs to be currently handled in spmd expansion.

Differential Revision: [D44615039](https://our.internmc.facebook.com/intern/diff/D44615039/)

[ghstack-poisoned]

Commit:
Update on "[spmd expansion] support scalar_tensor"


scalar_tensor is a pure factory function that can't be handled by DTensor prop rule and needs to be currently handled in spmd expansion.

Differential Revision: [D44615039](https://our.internmc.facebook.com/intern/diff/D44615039/)

[ghstack-poisoned]

yifuwang commented:
@yifuwang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44615039).

yifuwang commented:
@yifuwang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44615039).

Commit:
Update on "[spmd expansion] support scalar_tensor"


scalar_tensor is a pure factory function that can't be handled by DTensor prop rule and needs to be currently handled in spmd expansion.

Differential Revision: [D44615039](https://our.internmc.facebook.com/intern/diff/D44615039/)

[ghstack-poisoned]

yifuwang commented:
@yifuwang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D44615039).

yifuwang commented:
Superceded by https://github.com/pytorch/pytorch/pull/98390
